lib/taste_tester/commands.rb

```diff
@@ -210,5 +210,165 @@ def self.upload
+      # Use the repository specified in config.rb to calculate the changes
+      # that may affect Chef. These changes will be further analyzed to
+      # determine specific roles which may change due to modifed dependencies.
+      repo = BetweenMeals::Repo.get(
+        TasteTester::Config.repo_type,
+        TasteTester::Config.repo,
+        logger,
+      )
+      if repo && !repo.exists?
+        fail "Could not open repo from #{TasteTester::Config.repo}"
+      end
+
+      changes = _find_changeset(repo)
+
+      # Use Knife (or custom logic) to check the dependencies of each role
+      # against the list of changes. `impacted_roles` will contian the set
+      # of roles with direct or indirect (dependency) modifications.
+      impacted_roles = TasteTester::Hooks.impact_find_roles(changes)
+      impacted_roles ||= _find_roles(changes)
+
+      # Do any post processing required on the list of impacted roles, such
+      # as looking up hostnames associated with each role.
+      final_impact = TasteTester::Hooks.post_impact(impacted_roles)
+      final_impact ||= impacted_roles
+
+      # Print the calculated impact. If a print hook is defined that
+      # returns true, then the default print function is skipped.
+      unless TasteTester::Hooks.print_impact(final_impact)
+        _print_impact(final_impact)
+      end
+    end
+
+    def self._find_changeset(repo)
+      # We want to compare changes in the current directory (working set) with
+      # the "most recent" commit in the VCS. For SVN, this will be the latest
+      # commit on the checked out repository (i.e. 'trunk'). Git/Hg may have
+      # different tags or labels assigned to the master branch, (i.e. 'master',
+      # 'stable', etc.) and should be configured if different than the default.
+      start_ref = case repo
+                  when BetweenMeals::Repo::Svn
+                    repo.latest_revision
+                  when BetweenMeals::Repo::Git
+                    TasteTester::Config.vcs_start_ref_git
+                  when BetweenMeals::Repo::Hg
+                    TasteTester::Config.vcs_start_ref_hg
+                  end
+      end_ref = TasteTester::Config.vcs_end_ref
+
+      changeset = BetweenMeals::Changeset.new(
+        logger,
+        repo,
+        start_ref,
+        end_ref,
+        {
+          :cookbook_dirs =>
+            TasteTester::Config.relative_cookbook_dirs,
+          :role_dir =>
+            TasteTester::Config.relative_role_dir,
+          :databag_dir =>
+            TasteTester::Config.relative_databag_dir,
+        },
+        @track_symlinks,
+      )
+
+      return changeset
+    end
+
+    def self._find_roles(changes)
+      if TasteTester::Config.relative_cookbook_dirs.length > 1
+        logger.error('Knife deps does not support multiple cookbook paths.')
+        logger.error('Please flatten the cookbooks into a single directory' +
+                     ' or override the impact_find_roles function.')
```

Comment:
" or define the impact_find_roles method in a local plugin"

Reply:
Hopefully that error message wont live long :)
(https://github.com/chef/chef/issues/8651)